### PR TITLE
[FLINK-9850] Add a string to the print method to identify output for DataStream

### PR DESCRIPTION
**Pull request by yanghua**
## What is the purpose of the change

*This pull request adds a string to the print method to identify output for DataStream*

## Brief change log

  - *add print(string) / printToErr(string) to DataStream Java API*
  - *add print(string) / printToErr(string) to DataStream Scala API*
  - *add print(string) to DataStream Python API*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by yanghua** [comment]
@tillrohrmann and @zentol I see the Python DataStream API methods do not match DataStream Java API methods (missed some API methods), Shall we add those missed API into `PythonDataStream`? If yes, I'd like to do this.

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala [hunk 1]
```diff
@@ -959,6 +959,29 @@ class DataStream[T](stream: JavaStream[T]) {
   @PublicEvolving
   def printToErr() = stream.printToErr()
 
+  /**
+    * Writes a DataStream to the standard output stream (stdout). For each
+    * element of the DataStream the result of .toString is
```
Comment (hequn8128) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala]: .toString => [[AnyRef.toString()]]

**Comment by yanghua** [comment]
@hequn8128  thanks, I have added some test case~

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/PrintSinkFunction.java [hunk 2]
```diff
@@ -40,6 +40,8 @@
 	private boolean target;
 	private transient PrintStream stream;
 	private transient String prefix;
+	private String sinkIdentifier;
+	private String completedPrefix;
```
Comment (tison1) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/PrintSinkFunction.java]: if `prefix` is `transient`, why not `completedPrefix`?

**Comment by hequn8128** [comment]
@yanghua Thanks for your update. +1 to merge

**Comment by tison1** [comment]
@yanghua also +1

this is a net win.

**Comment by yanghua** [comment]
cc @tillrohrmann and @zentol

**Comment by yanghua** [comment]
cc @dawidwys and @pnowojski

**Comment by yanghua** [comment]
cc @tillrohrmann and @zentol

**Comment by yanghua** [comment]
cc @tillrohrmann

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/PrintSinkFunctionTest.java [hunk 3]
```diff
@@ -117,6 +117,65 @@ public void testPrintSinkWithPrefix() throws Exception {
...
+		assertEquals("Print to System.err", printSink.toString());
+		assertEquals("mySink:2> hello world!" + line, baos.toString());
+
+		printSink.close();
+		stream.close();
+	}
+
+	@Test
+	public void testPrintSinkWithIdentifierButNoPrefix() throws Exception {
+		ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/PrintSinkFunctionTest.java]: There is quite a lot of code duplication in those tests and they were unnecessarily using mockito instead of proper mock. Also there was even a bug in `testPrintSinkStdErr`. I have fixed those issues in a hotifx: https://github.com/apache/flink/pull/6538

I would the hotifx to be merged before this PR and please adapt/rewrite your test in similar fashion as I did in my hotfix.

**Comment by yanghua** [comment]
Hi @pnowojski thanks for your suggestion, I have refactored this PR, we both changed the `PrintSinkFunctionTest` so maybe I did not finish all of the works. What's more, I agree with the opinion about "deduplicate the logic of PrintSinkFunction and PrintingOutputFormat",  I'd like to start this work, if current issue would be fixed.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/PrintSinkFunction.java [hunk 4]
```diff
@@ -70,27 +84,30 @@ public void open(Configuration parameters) throws Exception {
...
+		if (!completedPrefix.isEmpty()) {
+			completedPrefix += "> ";
+		}
 	}
 
 	@Override
 	public void invoke(IN record) {
-		if (prefix != null) {
-			stream.println(prefix + record.toString());
+		if (completedPrefix != null) {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/PrintSinkFunction.java]: This check is not valid anymore, this field is never null.

**Comment by yanghua** [comment]
@pnowojski fixed some issues you just mentioned, will refactor `PrintSinkFunction ` and `PrintingOutputFormat `, you mean I should create a new class named `TaskOutputWriter ` and extract the same logic code?

**Comment by yanghua** [comment]
cc @pnowojski updated

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		if (!completedPrefix.isEmpty()) {
+			completedPrefix += "> ";
+		}
+	}
+
+	public void write(IN record) {
+		stream.println(completedPrefix + record.toString());
+	}
+
+	public void close() {
```
Comment (pnowojski) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java]: Drop the close? It seems like a dead empty method.

File: flink-java/src/test/java/org/apache/flink/api/java/io/PrintingOutputFormatTest.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests for the {@link PrintingOutputFormat}.
+ */
+public class PrintingOutputFormatTest {
```
Comment (pnowojski) [inline flink-java/src/test/java/org/apache/flink/api/java/io/PrintingOutputFormatTest.java]: +1 for adding those tests!

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/PrintSinkFunction.java [hunk 7]
```diff
@@ -18,81 +18,74 @@
...
+		writer = new TaskOutputWriter<>(sinkIdentifier, stdErr);
 	}
 
 	@Override
 	public void open(Configuration parameters) throws Exception {
 		super.open(parameters);
 		StreamingRuntimeContext context = (StreamingRuntimeContext) getRuntimeContext();
-		// get the target stream
-		stream = target == STD_OUT ? System.out : System.err;
+		int taskNumber = context.getIndexOfThisSubtask();
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/PrintSinkFunction.java]: nit: inline those variables?

File: flink-java/src/test/java/org/apache/flink/api/java/io/PrintingOutputFormatTest.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+		printSink.open(1, 2);
+
+		printSink.writeRecord("hello world!");
+
+		assertEquals("Print to System.out", printSink.toString());
+		assertEquals("mySink:2> hello world!" + line, arrayOutputStream.toString());
+
+		printSink = new PrintingOutputFormat<>("mySink", true);
+		assertEquals("Print to System.err", printSink.toString());
+		assertEquals("mySink:2> hello world!" + line, arrayOutputStream.toString());
```
Comment (pnowojski) [inline flink-java/src/test/java/org/apache/flink/api/java/io/PrintingOutputFormatTest.java]: Were you trying in those 3 lines test for stderr? If so, you haven't invoked anything on stderr instance of `printSink` and in this line, you are just asserting again the standard output. For asserting stderr you should check against `arrayErrorStream` instead of `arrayOutputStream`. 

However, I don't think it is necessary here to test for prefix AND for stderr together. So I would be actually in favour of dropping those 3 lines here (If you want you can keep those assertions after fixing the before mentioned bug).

File: flink-java/src/test/java/org/apache/flink/api/java/io/PrintingOutputFormatTest.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+	public void testPrintOutputFormatWithIdentifierAndPrefix() throws Exception {
+		PrintingOutputFormat<String> printSink = new PrintingOutputFormat<>("mySink", false);
+		printSink.open(1, 2);
+
+		printSink.writeRecord("hello world!");
+
+		assertEquals("Print to System.out", printSink.toString());
+		assertEquals("mySink:2> hello world!" + line, arrayOutputStream.toString());
+
+		printSink = new PrintingOutputFormat<>("mySink", true);
```
Comment (pnowojski) [inline flink-java/src/test/java/org/apache/flink/api/java/io/PrintingOutputFormatTest.java]: nit: you haven't closed the previous sink.

File: flink-java/src/test/java/org/apache/flink/api/java/io/PrintingOutputFormatTest.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+		printSink.open(0, 1);
+
+		printSink.writeRecord("hello world!");
+
+		assertEquals("Print to System.out", printSink.toString());
+		assertEquals("mySink> hello world!" + line, arrayOutputStream.toString());
+
+		printSink = new PrintingOutputFormat<>("mySink", true);
+		assertEquals("Print to System.err", printSink.toString());
+		assertEquals("mySink> hello world!" + line, arrayOutputStream.toString());
```
Comment (pnowojski) [inline flink-java/src/test/java/org/apache/flink/api/java/io/PrintingOutputFormatTest.java]: ditto regarding testing stderr.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/PrintSinkFunctionTest.java [hunk 11]
```diff
@@ -103,4 +102,41 @@ public void testPrintSinkWithPrefix() throws Exception {
...
+		printSink.open(new Configuration());
+
+		printSink.invoke("hello world!", SinkContextUtil.forTimestamp(0));
+
+		assertEquals("Print to System.out", printSink.toString());
+		assertEquals("mySink:2> hello world!" + line, arrayOutputStream.toString());
+
+		printSink = new PrintSinkFunction<>("mySink", true);
+		assertEquals("Print to System.err", printSink.toString());
+		assertEquals("mySink:2> hello world!" + line, arrayOutputStream.toString());
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/PrintSinkFunctionTest.java]: ditto regarding testing stderr

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/PrintSinkFunctionTest.java [hunk 12]
```diff
@@ -103,4 +102,41 @@ public void testPrintSinkWithPrefix() throws Exception {
...
+		printSink.open(new Configuration());
+
+		printSink.invoke("hello world!", SinkContextUtil.forTimestamp(0));
+
+		assertEquals("Print to System.out", printSink.toString());
+		assertEquals("mySink> hello world!" + line, arrayOutputStream.toString());
+
+		printSink = new PrintSinkFunction<>("mySink", true);
+		assertEquals("Print to System.err", printSink.toString());
+		assertEquals("mySink> hello world!" + line, arrayOutputStream.toString());
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/PrintSinkFunctionTest.java]: ditto regarding testing stderr

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/MockRuntimeContext.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.flink.runtime.operators.testutils.MockEnvironmentBuilder;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.StreamingRuntimeContext;
+
+import java.util.Collections;
+
+/**
+ * Simple test runtime context for stream operators.
+ */
+public class MockRuntimeContext extends StreamingRuntimeContext {
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/util/MockRuntimeContext.java]: Why did you add this class? I have already extracted the same thing on the master branch to `MockStreamingRuntimeContext`. I'm guessing some rebase error?

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+package org.apache.flink.api.common.functions.util;
+
+import java.io.PrintStream;
+import java.io.Serializable;
+
+/**
+ * Task output writer for DataStream and DataSet print API.
+ */
+public class TaskOutputWriter<IN> implements Serializable {
```
Comment (pnowojski) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java]: add `@Internal`?

Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java]: can we give this a different name, like `PrintSinkOutputWriter`? The current name is a bit too general imo.

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+	public TaskOutputWriter(final boolean stdErr) {
+		this("", stdErr);
+	}
+
+	public TaskOutputWriter(final String sinkIdentifier, final boolean stdErr) {
+		this.target = stdErr;
+		this.sinkIdentifier = (sinkIdentifier == null ? "" : sinkIdentifier);
+	}
+
+	public void open(int taskNumber, int numTasks) {
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java]: rename arguments to `subtaskIndex` and `numParallelSubtasks`

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+	public TaskOutputWriter() {
+		this("", STD_OUT);
+	}
+
+	public TaskOutputWriter(final boolean stdErr) {
+		this("", stdErr);
+	}
+
+	public TaskOutputWriter(final String sinkIdentifier, final boolean stdErr) {
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java]: This class could be simplified by simply passing in the `OutputStream`.

File: flink-libraries/flink-streaming-python/src/main/java/org/apache/flink/streaming/python/api/datastream/PythonDataStream.java [hunk 17]
```diff
@@ -144,6 +144,14 @@ public void output() {
 		stream.print();
 	}
 
+	/**
+	 * A thin wrapper layer over {@link DataStream#print(String)}.
+	 */
+	@PublicEvolving
+	public void output(String sinkIdentifier) {
+		stream.print(sinkIdentifier);
```
Comment (zentol) [inline flink-libraries/flink-streaming-python/src/main/java/org/apache/flink/streaming/python/api/datastream/PythonDataStream.java]: this is not tested

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/PrintSinkFunction.java [hunk 18]
```diff
@@ -18,81 +18,74 @@
...
-	private transient PrintStream stream;
-	private transient String prefix;
+	private final TaskOutputWriter<IN> writer;
 
 	/**
 	 * Instantiates a print sink function that prints to standard out.
 	 */
-	public PrintSinkFunction() {}
+	public PrintSinkFunction() {
+		writer = new TaskOutputWriter<>();
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/PrintSinkFunction.java]: call the other constructor instead

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala [hunk 19]
```diff
@@ -959,6 +959,29 @@ class DataStream[T](stream: JavaStream[T]) {
...
+  /**
+    * Writes a DataStream to the standard output stream (stdout). For each
+    * element of the DataStream the result of [[AnyRef.toString()]] is
+    * written.
+    *
+    * @param sinkIdentifier The string to prefix the output with.
+    * @return The closed DataStream.
+    */
+  @PublicEvolving
+  def print(sinkIdentifier: String): DataStreamSink[T] = stream.print(sinkIdentifier)
```
Comment (zentol) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala]: these are not tested

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/MockRuntimeContext.java (hunk 13 above)
Comment (yanghua) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/util/MockRuntimeContext.java]: This is based on your previous suggestion. Before your PR was merged, I forgot to delete it. Sorry, it will be deleted.

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java (hunk 15 above)
Comment (yanghua) [reply flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java]: This suggestion is good, shall we rename the arguments of the interface method `OutputFormat#open`, or can we rename them in another issue? or let it go?

**Comment by yanghua** [comment]
cc @pnowojski

File: flink-libraries/flink-streaming-python/src/test/java/org/apache/flink/streaming/python/api/PythonStreamBinderTest.java [hunk 20]
```diff
@@ -63,7 +63,7 @@ private static Path findUtilsModule() {
 
 	@Test
 	public void testProgram() throws Exception {
-		Path testEntryPoint = new Path(getBaseTestPythonDir(), "examples/word_count.py");
+		Path testEntryPoint = new Path(getBaseTestPythonDir(), "run_all_tests.py");
```
Comment (pnowojski) [inline flink-libraries/flink-streaming-python/src/test/java/org/apache/flink/streaming/python/api/PythonStreamBinderTest.java]: Is this change relevant?

File: flink-libraries/flink-streaming-python/src/test/python/org/apache/flink/streaming/python/api/test_output_with_sink_identifier.py [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+        env = flink.get_execution_environment()
+        env.from_collection(elements) \
+            .flat_map(Tokenizer()) \
+            .key_by(Selector()) \
+            .time_window(milliseconds(10)) \
+            .reduce(Sum()) \
+            .output("testOutputWithIdentifier")
+
+        env.execute()
```
Comment (pnowojski) [inline flink-libraries/flink-streaming-python/src/test/python/org/apache/flink/streaming/python/api/test_output_with_sink_identifier.py]: This is first time I'm looking at our python tests, so take my comment with a grain of salt.

Isn't this test (and other python streaming tests) supposed to assert on something? Aren't they now testing only that python api doesn't throw anything but it would still pass with empty methods? @zentol  ?

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java (hunk 15 above)
Comment (zentol) [reply flink-core/src/main/java/org/apache/flink/api/common/functions/util/TaskOutputWriter.java]: ignore the `OutputFormat#open` method,

File: flink-libraries/flink-streaming-python/src/test/python/org/apache/flink/streaming/python/api/test_output_with_sink_identifier.py (hunk 21 above)
Comment (zentol) [reply flink-libraries/flink-streaming-python/src/test/python/org/apache/flink/streaming/python/api/test_output_with_sink_identifier.py]: yes this is a common issue with the tests, but fixing this isn't too high on my priority list, given that most methods are just wrappers around java methods.

File: flink-libraries/flink-streaming-python/src/test/java/org/apache/flink/streaming/python/api/PythonStreamBinderTest.java (hunk 20 above)
Comment (zentol) [reply flink-libraries/flink-streaming-python/src/test/java/org/apache/flink/streaming/python/api/PythonStreamBinderTest.java]: Basically this change actually enables most python tests, which i accidentally disabled months ago. So yeah, to enable more tests we do need this.

**Comment by zentol** [comment]
Personally I would prefer if we'd not change the python API in this PR, because we're now in the exact position we don't want to be in: where the merging of the PR is delayed due to another effectively unrelated problem.

**Comment by yanghua** [comment]
@zentol Do you mean that we don't change the python API here and remove the tests for the Python API?

**Comment by zentol** [comment]
yes.

**Comment by yanghua** [comment]
@pnowojski Python-related changes have been removed, only changes to Java and Scala

**Comment by pnowojski** [comment]
Thanks for the contribution! @zentol told me his LGTM so merging now :)
